File: building_monitor.py
"""
Трекер зданий для отслеживания улучшений - премиум функция
"""
import asyncio
import logging
import json
from datetime import datetime, timedelta
from typing import Dict, Any, List, Optional

from database import DatabaseService
from coc_api import CocApiClient
from models.building import BuildingSnapshot, BuildingUpgrade, BuildingTracker
from config import config

logger = logging.getLogger(__name__)


class BuildingMonitor:
    """Сервис мониторинга улучшений зданий для премиум пользователей"""
    
    def __init__(self, db_service: DatabaseService, coc_client: CocApiClient, bot_instance=None):
        self.db_service = db_service
        self.coc_client = coc_client
        self.bot = bot_instance
        
        self.is_running = False
        self.task = None
        
        # Интервал проверки - каждые 5 минут
        self.check_interval = 300  # 5 минут в секундах
        
        # Словарь для перевода названий зданий на русский
        self.building_names_ru = {
            "Town Hall": "Ратуша",
            "Army Camp": "Казарма",
            "Barracks": "Учебные казармы",
            "Laboratory": "Лаборатория",
            "Spell Factory": "Фабрика заклинаний",
            "Clan Castle": "Замок клана",
            "Gold Mine": "Золотая шахта",
            "Elixir Collector": "Накопитель эликсира",
            "Dark Elixir Drill": "Бур темного эликсира",
            "Gold Storage": "Хранилище золота",
            "Elixir Storage": "Хранилище эликсира",
            "Dark Elixir Storage": "Хранилище темного эликсира",
            "Cannon": "Пушка",
            "Archer Tower": "Башня лучниц",
            "Mortar": "Мортира",
            "Air Defense": "Воздушная защита",
            "Wizard Tower": "Башня магов",
            "Air Sweeper": "Воздушная метла",
            "Hidden Tesla": "Скрытая тесла",
            "Bomb Tower": "Башня-бомба",
            "X-Bow": "Адский лук",
            "Inferno Tower": "Башня ада",
            "Eagle Artillery": "Орлиная артиллерия",
            "Scattershot": "Разброс",
            "Builder's Hut": "Хижина строителя",
            "Barbarian King": "Король варваров",
            "Archer Queen": "Королева лучниц",
            "Grand Warden": "Великий хранитель",
            "Royal Champion": "Королевский чемпион"
        }
    
    async def start(self):
        """Запуск мониторинга зданий"""
        if self.is_running:
            logger.warning("Монитор зданий уже запущен")
            return
        
        self.is_running = True
        self.task = asyncio.create_task(self._monitoring_loop())
        logger.info("Сервис мониторинга зданий запущен")
    
    async def stop(self):
        """Остановка мониторинга зданий"""
        self.is_running = False
        if self.task:
            self.task.cancel()
            try:
                await self.task
            except asyncio.CancelledError:
                pass
        logger.info("Сервис мониторинга зданий остановлен")
    
    async def _monitoring_loop(self):
        """Основной цикл мониторинга"""
        while self.is_running:
            try:
                await self._check_all_trackers()
                
                # Ждем до следующей проверки (5 минут)
                await asyncio.sleep(self.check_interval)
                
            except asyncio.CancelledError:
                break
            except Exception as e:
                logger.error(f"[Монитор зданий] Ошибка в фоновой задаче: {e}")
                await asyncio.sleep(60)  # Ждем минуту перед повтором при ошибке
    
    async def _check_all_trackers(self):
        """Проверка всех активных отслеживателей"""
        try:
            trackers = await self.db_service.get_active_building_trackers()
            logger.info(f"[Монитор зданий] Проверка {len(trackers)} активных отслеживателей")
            
            for tracker in trackers:
                # Проверяем, что у пользователя есть активная премиум подписка
                subscription = await self.db_service.get_subscription(tracker.telegram_id)
                if not subscription or not subscription.is_active or subscription.is_expired():
                    logger.info(f"Отключение отслеживания для пользователя {tracker.telegram_id} - нет активной подписки")
                    await self._deactivate_tracker(tracker.telegram_id)
                    continue
                
                await self._check_player_buildings(tracker)
                
        except Exception as e:
            logger.error(f"[Монитор зданий] Ошибка при проверке отслеживателей: {e}")
    
    async def _check_player_buildings(self, tracker: BuildingTracker):
        """Проверка зданий конкретного игрока"""
        try:
            # Получаем текущую информацию о игроке
            async with self.coc_client as client:
                player_data = await client.get_player_info(tracker.player_tag)
                
                if not player_data:
                    logger.warning(f"Не удалось получить данные игрока {tracker.player_tag}")
                    return
            
            # Получаем последний снимок зданий
            last_snapshot = await self.db_service.get_latest_building_snapshot(tracker.player_tag)
            
            if not last_snapshot:
                # Создаем первый снимок
                await self._create_initial_snapshot(tracker.player_tag, player_data)
                logger.info(f"Создан первый снимок зданий для игрока {tracker.player_tag}")
                return
            
            # Сравниваем здания
            upgrades = await self._compare_buildings(last_snapshot, player_data)
            
            if upgrades:
                # Отправляем уведомления об улучшениях
                await self._send_upgrade_notifications(tracker.telegram_id, upgrades)
                
                # Сохраняем новый снимок
                await self._create_snapshot(tracker.player_tag, player_data)
            
            # Обновляем время последней проверки
            now = datetime.now().isoformat()
            await self.db_service.update_tracker_last_check(tracker.telegram_id, now)
            
        except Exception as e:
            logger.error(f"[Монитор зданий] Ошибка при проверке игрока {tracker.player_tag}: {e}")
    
    async def _create_initial_snapshot(self, player_tag: str, player_data: Dict[Any, Any]):
        """Создание первого снимка зданий"""
        await self._create_snapshot(player_tag, player_data)
    
    async def _create_snapshot(self, player_tag: str, player_data: Dict[Any, Any]):
        """Создание снимка состояния зданий"""
        try:
            buildings_data = {}
            
            # Извлекаем данные о зданиях (включая ратушу)
            if 'townHallLevel' in player_data:
                buildings_data['Town Hall'] = player_data['townHallLevel']
            
            # Извлекаем данные о героях
            if 'heroes' in player_data:
                for hero in player_data['heroes']:
                    if 'name' in hero and 'level' in hero:
                        buildings_data[hero['name']] = hero['level']
            
            # Извлекаем данные о войсках (войска могут улучшаться в лаборатории)
            if 'troops' in player_data:
                for troop in player_data['troops']:
                    if 'name' in troop and 'level' in troop:
                        buildings_data[f"{troop['name']} (войска)"] = troop['level']
            
            # Извлекаем данные о заклинаниях
            if 'spells' in player_data:
                for spell in player_data['spells']:
                    if 'name' in spell and 'level' in spell:
                        buildings_data[f"{spell['name']} (заклинание)"] = spell['level']
            
            snapshot = BuildingSnapshot(
                player_tag=player_tag,
                snapshot_time=datetime.now().isoformat(),
                buildings_data=json.dumps(buildings_data)
            )
            
            await self.db_service.save_building_snapshot(snapshot)
            
        except Exception as e:
            logger.error(f"Ошибка при создании снимка зданий: {e}")
    
    async def _compare_buildings(self, last_snapshot: BuildingSnapshot, current_data: Dict[Any, Any]) -> List[BuildingUpgrade]:
        """Сравнение зданий и поиск улучшений"""
        upgrades = []
        
        try:
            # Загружаем данные из последнего снимка
            old_buildings = json.loads(last_snapshot.buildings_data)
            
            # Создаем текущий снимок для сравнения
            current_buildings = {}
            
            # Ратуша
            if 'townHallLevel' in current_data:
                current_buildings['Town Hall'] = current_data['townHallLevel']
            
            # Герои
            if 'heroes' in current_data:
                for hero in current_data['heroes']:
                    if 'name' in hero and 'level' in hero:
                        current_buildings[hero['name']] = hero['level']
            
            # Войска
            if 'troops' in current_data:
                for troop in current_data['troops']:
                    if 'name' in troop and 'level' in troop:
                        current_buildings[f"{troop['name']} (войска)"] = troop['level']
            
            # Заклинания
            if 'spells' in current_data:
                for spell in current_data['spells']:
                    if 'name' in spell and 'level' in spell:
                        current_buildings[f"{spell['name']} (заклинание)"] = spell['level']
            
            # Сравниваем уровни
            for building_name, current_level in current_buildings.items():
                old_level = old_buildings.get(building_name, 0)
                
                if current_level > old_level:
                    upgrades.append(BuildingUpgrade(
                        building_name=building_name,
                        old_level=old_level,
                        new_level=current_level
                    ))
            
        except Exception as e:
            logger.error(f"Ошибка при сравнении зданий: {e}")
        
        return upgrades
    
    async def _send_upgrade_notifications(self, telegram_id: int, upgrades: List[BuildingUpgrade]):
        """Отправка уведомлений об улучшениях"""
        if not self.bot:
            return
        
        try:
            for upgrade in upgrades:
                # Переводим название на русский
                building_name_ru = self.building_names_ru.get(upgrade.building_name, upgrade.building_name)
                
                message = (
                    f"🏗️ <b>Улучшение завершено!</b>\n\n"
                    f"🔨 {building_name_ru} улучшен с {upgrade.old_level} на {upgrade.new_level} уровень!\n\n"
                    f"🎉 Поздравляем с успешным улучшением!"
                )
                
                await self.bot.send_message(
                    chat_id=telegram_id,
                    text=message,
                    parse_mode='HTML'
                )
                
                logger.info(f"Отправлено уведомление об улучшении {building_name_ru} пользователю {telegram_id}")
                
        except Exception as e:
            logger.error(f"Ошибка при отправке уведомлений: {e}")
    
    async def activate_tracking(self, telegram_id: int, player_tag: str) -> bool:
        """Активация отслеживания зданий для пользователя"""
        try:
            # Проверяем, что у пользователя есть премиум подписка
            subscription = await self.db_service.get_subscription(telegram_id)
            if not subscription or not subscription.is_active or subscription.is_expired():
                return False
            
            tracker = BuildingTracker(
                telegram_id=telegram_id,
                player_tag=player_tag,
                is_active=True,
                created_at=datetime.now().isoformat()
            )
            
            success = await self.db_service.save_building_tracker(tracker)
            if success:
                logger.info(f"Активировано отслеживание зданий для пользователя {telegram_id}, игрок {player_tag}")
            
            return success
            
        except Exception as e:
            logger.error(f"Ошибка при активации отслеживания: {e}")
            return False
    
    async def _deactivate_tracker(self, telegram_id: int):
        """Деактивация отслеживания"""
        try:
            tracker = await self.db_service.get_building_tracker(telegram_id)
            if tracker:
                tracker.is_active = False
                await self.db_service.save_building_tracker(tracker)
                logger.info(f"Отслеживание зданий деактивировано для пользователя {telegram_id}")
                
        except Exception as e:
            logger.error(f"Ошибка при деактивации отслеживания: {e}")
    
    async def deactivate_tracking(self, telegram_id: int) -> bool:
        """Деактивация отслеживания зданий пользователем"""
        try:
            await self._deactivate_tracker(telegram_id)
            return True
        except Exception as e:
            logger.error(f"Ошибка при деактивации отслеживания пользователем: {e}")
            return False
    
    async def is_tracking_active(self, telegram_id: int) -> bool:
        """Проверка активности отслеживания для пользователя"""
        try:
            tracker = await self.db_service.get_building_tracker(telegram_id)
            return tracker is not None and tracker.is_active
        except Exception as e:
            logger.error(f"Ошибка при проверке статуса отслеживания: {e}")
            return False

File: internal/services/building_monitor.go
// Package services предоставляет бизнес-логику и сервисы бота
package services

import (
	"context"
	"encoding/json"
	"log"
	"sync"
	"time"

	"ClashBOfClashBot/internal/api"
	"ClashBOfClashBot/internal/database"
	"ClashBOfClashBot/internal/models"
)

// BuildingMonitor - сервис мониторинга улучшений зданий для премиум пользователей
type BuildingMonitor struct {
	dbService        *database.DatabaseService
	cocClient        *api.CocApiClient
	botInstance      interface{} // Telegram bot instance
	isRunning        bool
	mutex            sync.RWMutex
	ctx              context.Context
	cancel           context.CancelFunc
	minCheckInterval time.Duration
	buildingNamesRU  map[string]string
}

// NewBuildingMonitor создает новый экземпляр BuildingMonitor
func NewBuildingMonitor(dbService *database.DatabaseService, cocClient *api.CocApiClient, botInstance interface{}) *BuildingMonitor {
	ctx, cancel := context.WithCancel(context.Background())

	return &BuildingMonitor{
		dbService:        dbService,
		cocClient:        cocClient,
		botInstance:      botInstance,
		isRunning:        false,
		ctx:              ctx,
		cancel:           cancel,
		minCheckInterval: 90 * time.Second, // 90 секунд (1.5 минуты) - интервал для всех пользователей
		buildingNamesRU:  getBuildingNamesRU(),
	}
}

// getBuildingNamesRU возвращает словарь для перевода названий зданий на русский
func getBuildingNamesRU() map[string]string {
	return map[string]string{
		"Town Hall":                    "Ратуша",
		"Army Camp":                    "Казарма",
		"Barracks":                     "Учебные казармы",
		"Laboratory":                   "Лаборатория",
		"Spell Factory":                "Фабрика заклинаний",
		"Clan Castle":                  "Замок клана",
		"Gold Mine":                    "Золотая шахта",
		"Elixir Collector":             "Накопитель эликсира",
		"Dark Elixir Drill":            "Бур темного эликсира",
		"Gold Storage":                 "Хранилище золота",
		"Elixir Storage":               "Хранилище эликсира",
		"Dark Elixir Storage":          "Хранилище темного эликсира",
		"Cannon":                       "Пушка",
		"Archer Tower":                 "Башня лучниц",
		"Mortar":                       "Мортира",
		"Air Defense":                  "Воздушная защита",
		"Wizard Tower":                 "Башня магов",
		"Air Sweeper":                  "Воздушная метла",
		"Hidden Tesla":                 "Скрытая тесла",
		"Bomb Tower":                   "Башня-бомба",
		"X-Bow":                        "Адский лук",
		"Inferno Tower":                "Башня ада",
		"Eagle Artillery":              "Орлиная артиллерия",
		"Scattershot":                  "Разброс",
		"Builder's Hut":                "Хижина строителя",
		"Barbarian King":               "Король варваров",
		"Archer Queen":                 "Королева лучниц",
		"Grand Warden":                 "Великий хранитель",
		"Royal Champion":               "Королевский чемпион",
		"Walls (стены)":                "Стены",
		"[БД] Builder Hall":            "Зал строителя",
		"[БД] League":                  "Лига деревни строителя",
	}
}

// Start запускает мониторинг зданий
func (b *BuildingMonitor) Start() error {
	b.mutex.Lock()
	defer b.mutex.Unlock()

	if b.isRunning {
		log.Println("[Монитор зданий] Уже запущен")
		return nil
	}

	b.isRunning = true
	log.Println("[Монитор зданий] Сервис мониторинга зданий запущен")

	// Запускаем основной цикл в горутине
	go b.monitoringLoop()

	return nil
}

// Stop останавливает мониторинг зданий
func (b *BuildingMonitor) Stop() error {
	b.mutex.Lock()
	defer b.mutex.Unlock()

	if !b.isRunning {
		log.Println("[Монитор зданий] Уже остановлен")
		return nil
	}

	b.isRunning = false
	b.cancel()
	log.Println("[Монитор зданий] Сервис мониторинга зданий остановлен")

	return nil
}

// monitoringLoop основной цикл мониторинга
func (b *BuildingMonitor) monitoringLoop() {
	ticker := time.NewTicker(b.minCheckInterval)
	defer ticker.Stop()

	for {
		select {
		case <-b.ctx.Done():
			return
		case <-ticker.C:
			if err := b.checkAllTrackers(); err != nil {
				log.Printf("[Монитор зданий] Ошибка в фоновой задаче: %v", err)
				time.Sleep(time.Minute) // Ждем минуту перед повтором при ошибке
			}
		}
	}
}

// checkAllTrackers проверяет все активные отслеживатели
func (b *BuildingMonitor) checkAllTrackers() error {
	trackers, err := b.dbService.GetActiveBuildingTrackers()
	if err != nil {
		return err
	}

	log.Printf("[Монитор зданий] Проверка %d активных отслеживателей", len(trackers))

	currentTime := time.Now()

	for _, tracker := range trackers {
		// Проверяем, что у пользователя есть активная подписка
		subscription, err := b.dbService.GetSubscription(tracker.TelegramID)
		if err != nil || subscription == nil || !subscription.IsActive() || subscription.IsExpired() {
			log.Printf("[Монитор зданий] Отключение отслеживания для пользователя %d - нет активной подписки", tracker.TelegramID)
			if err := b.deactivateTracker(tracker.TelegramID); err != nil {
				log.Printf("[Монитор зданий] Ошибка при деактивации трекера: %v", err)
			}
			continue
		}

		// Определяем интервал проверки для данного пользователя
		checkInterval := b.getCheckIntervalForSubscription(subscription.SubscriptionType)

		// Проверяем, прошло ли достаточно времени с последней проверки
		if tracker.LastCheck != "" {
			lastCheckTime, err := time.Parse(time.RFC3339, tracker.LastCheck)
			if err == nil {
				timeSinceLastCheck := currentTime.Sub(lastCheckTime)
				if timeSinceLastCheck < checkInterval {
					continue // Ещё не время проверять этого пользователя
				}
			}
		}

		if err := b.checkPlayerBuildings(&tracker); err != nil {
			log.Printf("[Монитор зданий] Ошибка при проверке игрока %s: %v", tracker.PlayerTag, err)
		}
	}

	return nil
}

// getCheckIntervalForSubscription определяет интервал проверки по типу подписки
func (b *BuildingMonitor) getCheckIntervalForSubscription(subscriptionType string) time.Duration {
	// Согласно политике фан контента SuperCell - для всех аккаунтов с включенными уведомлениями
	// проверка улучшений строений происходит раз в 1.5 минуты (90 секунд)
	return 90 * time.Second // 90 секунд (1.5 минуты) для всех пользователей
}

// checkPlayerBuildings проверяет здания конкретного игрока
func (b *BuildingMonitor) checkPlayerBuildings(tracker *models.BuildingTracker) error {
	// Получаем текущую информацию о игроке
	playerData, err := b.cocClient.GetPlayerInfo(tracker.PlayerTag)
	if err != nil {
		return err
	}

	if playerData == nil {
		log.Printf("[Монитор зданий] Не удалось получить данные игрока %s", tracker.PlayerTag)
		return nil
	}

	// Получаем последний снимок зданий
	lastSnapshot, err := b.dbService.GetLatestBuildingSnapshot(tracker.PlayerTag)
	if err != nil {
		return err
	}

	if lastSnapshot == nil {
		// Создаем первый снимок
		if err := b.createSnapshot(tracker.PlayerTag, playerData); err != nil {
			return err
		}
		log.Printf("[Монитор зданий] Создан первый снимок зданий для игрока %s", tracker.PlayerTag)
		return nil
	}

	// Сравниваем здания
	upgrades, err := b.compareBuildings(lastSnapshot, playerData)
	if err != nil {
		return err
	}

	if len(upgrades) > 0 {
		// Отправляем уведомления об улучшениях
		if err := b.sendUpgradeNotifications(tracker.TelegramID, upgrades, tracker.PlayerTag); err != nil {
			log.Printf("[Монитор зданий] Ошибка при отправке уведомлений: %v", err)
		}

		// Сохраняем новый снимок
		if err := b.createSnapshot(tracker.PlayerTag, playerData); err != nil {
			return err
		}
	}

	// Обновляем время последней проверки
	now := time.Now().Format(time.RFC3339)
	if err := b.dbService.UpdateTrackerLastCheck(tracker.TelegramID, now, tracker.PlayerTag); err != nil {
		return err
	}

	return nil
}

// createSnapshot создает снимок состояния зданий
func (b *BuildingMonitor) createSnapshot(playerTag string, playerData map[string]interface{}) error {
	buildingsData := make(map[string]interface{})

	// Извлекаем данные о зданиях (включая ратушу)
	if townHallLevel, ok := playerData["townHallLevel"].(float64); ok {
		buildingsData["Town Hall"] = int(townHallLevel)
	}

	// Извлекаем данные о героях
	if heroes, ok := playerData["heroes"].([]interface{}); ok {
		for _, h := range heroes {
			hero, ok := h.(map[string]interface{})
			if !ok {
				continue
			}

			name, nameOk := hero["name"].(string)
			level, levelOk := hero["level"].(float64)
			if nameOk && levelOk {
				buildingsData[name] = int(level)
			}
		}
	}

	// Извлекаем данные о снаряжении героев
	if heroEquipment, ok := playerData["heroEquipment"].([]interface{}); ok {
		for _, e := range heroEquipment {
			equipment, ok := e.(map[string]interface{})
			if !ok {
				continue
			}

			name, nameOk := equipment["name"].(string)
			level, levelOk := equipment["level"].(float64)
			if nameOk && levelOk {
				buildingsData[name+" (снаряжение)"] = int(level)
			}
		}
	}

	// Извлекаем данные о войсках
	if troops, ok := playerData["troops"].([]interface{}); ok {
		for _, t := range troops {
			troop, ok := t.(map[string]interface{})
			if !ok {
				continue
			}

			name, nameOk := troop["name"].(string)
			level, levelOk := troop["level"].(float64)
			if nameOk && levelOk {
				buildingsData[name+" (войска)"] = int(level)
			}
		}
	}

	// Извлекаем данные о заклинаниях
	if spells, ok := playerData["spells"].([]interface{}); ok {
		for _, s := range spells {
			spell, ok := s.(map[string]interface{})
			if !ok {
				continue
			}

			name, nameOk := spell["name"].(string)
			level, levelOk := spell["level"].(float64)
			if nameOk && levelOk {
				buildingsData[name+" (заклинание)"] = int(level)
			}
		}
	}

	// Извлекаем данные о стенах
	if achievements, ok := playerData["achievements"].([]interface{}); ok {
		for _, a := range achievements {
			achievement, ok := a.(map[string]interface{})
			if !ok {
				continue
			}

			name, nameOk := achievement["name"].(string)
			value, valueOk := achievement["value"].(float64)
			if nameOk && valueOk && name == "Wall Buster" {
				buildingsData["Walls (стены)"] = int(value)
				break
			}
		}
	}

	// Извлекаем данные о деревне строителя
	if builderHallLevel, ok := playerData["builderHallLevel"].(float64); ok {
		buildingsData["[БД] Builder Hall"] = int(builderHallLevel)
	}

	// Другие здания деревни строителя
	if builderBaseLeague, ok := playerData["builderBaseLeague"].(map[string]interface{}); ok {
		if leagueName, ok := builderBaseLeague["name"].(string); ok && leagueName != "Unranked" {
			buildingsData["[БД] League"] = leagueName
		}
	}

	// Сериализуем данные
	buildingsJSON, err := json.Marshal(buildingsData)
	if err != nil {
		return err
	}

	snapshot := models.NewBuildingSnapshot(
		playerTag,
		time.Now().Format(time.RFC3339),
		string(buildingsJSON),
	)

	return b.dbService.SaveBuildingSnapshot(snapshot)
}

// compareBuildings сравнивает здания и находит улучшения
func (b *BuildingMonitor) compareBuildings(lastSnapshot *models.BuildingSnapshot, currentData map[string]interface{}) ([]models.BuildingUpgrade, error) {
	upgrades := []models.BuildingUpgrade{}

	// Загружаем данные из последнего снимка
	var oldBuildings map[string]interface{}
	if err := json.Unmarshal([]byte(lastSnapshot.BuildingsData), &oldBuildings); err != nil {
		return nil, err
	}

	// Создаем текущий снимок для сравнения
	currentBuildings := make(map[string]interface{})

	// Ратуша
	if townHallLevel, ok := currentData["townHallLevel"].(float64); ok {
		currentBuildings["Town Hall"] = int(townHallLevel)
	}

	// Герои
	if heroes, ok := currentData["heroes"].([]interface{}); ok {
		for _, h := range heroes {
			hero, ok := h.(map[string]interface{})
			if !ok {
				continue
			}

			name, nameOk := hero["name"].(string)
			level, levelOk := hero["level"].(float64)
			if nameOk && levelOk {
				currentBuildings[name] = int(level)
			}
		}
	}

	// Снаряжение героев
	if heroEquipment, ok := currentData["heroEquipment"].([]interface{}); ok {
		for _, e := range heroEquipment {
			equipment, ok := e.(map[string]interface{})
			if !ok {
				continue
			}

			name, nameOk := equipment["name"].(string)
			level, levelOk := equipment["level"].(float64)
			if nameOk && levelOk {
				currentBuildings[name+" (снаряжение)"] = int(level)
			}
		}
	}

	// Войска
	if troops, ok := currentData["troops"].([]interface{}); ok {
		for _, t := range troops {
			troop, ok := t.(map[string]interface{})
			if !ok {
				continue
			}

			name, nameOk := troop["name"].(string)
			level, levelOk := troop["level"].(float64)
			if nameOk && levelOk {
				currentBuildings[name+" (войска)"] = int(level)
			}
		}
	}

	// Заклинания
	if spells, ok := currentData["spells"].([]interface{}); ok {
		for _, s := range spells {
			spell, ok := s.(map[string]interface{})
			if !ok {
				continue
			}

			name, nameOk := spell["name"].(string)
			level, levelOk := spell["level"].(float64)
			if nameOk && levelOk {
				currentBuildings[name+" (заклинание)"] = int(level)
			}
		}
	}

	// Стены
	if achievements, ok := currentData["achievements"].([]interface{}); ok {
		for _, a := range achievements {
			achievement, ok := a.(map[string]interface{})
			if !ok {
				continue
			}

			name, nameOk := achievement["name"].(string)
			value, valueOk := achievement["value"].(float64)
			if nameOk && valueOk && name == "Wall Buster" {
				currentBuildings["Walls (стены)"] = int(value)
				break
			}
		}
	}

	// Деревня строителя
	if builderHallLevel, ok := currentData["builderHallLevel"].(float64); ok {
		currentBuildings["[БД] Builder Hall"] = int(builderHallLevel)
	}

	if builderBaseLeague, ok := currentData["builderBaseLeague"].(map[string]interface{}); ok {
		if leagueName, ok := builderBaseLeague["name"].(string); ok && leagueName != "Unranked" {
			currentBuildings["[БД] League"] = leagueName
		}
	}

	// Сравниваем здания
	for buildingName, currentValue := range currentBuildings {
		oldValue, exists := oldBuildings[buildingName]

		// Преобразуем значения в int для сравнения
		var currentLevel, oldLevel int

		switch v := currentValue.(type) {
		case int:
			currentLevel = v
		case float64:
			currentLevel = int(v)
		case string:
			// Для строковых значений (например, лига) пропускаем
			continue
		}

		if exists {
			switch v := oldValue.(type) {
			case int:
				oldLevel = v
			case float64:
				oldLevel = int(v)
			case string:
				continue
			}

			// Проверяем, было ли улучшение
			if currentLevel > oldLevel {
				// Переводим название на русский
				ruName := buildingName
				if translated, ok := b.buildingNamesRU[buildingName]; ok {
					ruName = translated
				}

				upgrade := models.BuildingUpgrade{
					BuildingName: ruName,
					OldLevel:     oldLevel,
					NewLevel:     currentLevel,
					UpgradeTime:  time.Now().Format(time.RFC3339),
				}
				upgrades = append(upgrades, upgrade)
			}
		} else {
			// Новое здание/войска/заклинание
			ruName := buildingName
			if translated, ok := b.buildingNamesRU[buildingName]; ok {
				ruName = translated
			}

			upgrade := models.BuildingUpgrade{
				BuildingName: ruName,
				OldLevel:     0,
				NewLevel:     currentLevel,
				UpgradeTime:  time.Now().Format(time.RFC3339),
			}
			upgrades = append(upgrades, upgrade)
		}
	}

	return upgrades, nil
}

// sendUpgradeNotifications отправляет уведомления об улучшениях
func (b *BuildingMonitor) sendUpgradeNotifications(telegramID int64, upgrades []models.BuildingUpgrade, playerTag string) error {
	if b.botInstance == nil {
		return nil
	}

	// Здесь должна быть логика отправки уведомлений через Telegram Bot
	// Но так как это интерфейс, реализация будет в основном боте

	log.Printf("[Монитор зданий] Отправка уведомлений пользователю %d о %d улучшениях", telegramID, len(upgrades))

	return nil
}

// deactivateTracker деактивирует трекер
func (b *BuildingMonitor) deactivateTracker(telegramID int64) error {
	// Здесь должна быть логика деактивации трекера
	// Реализация будет зависеть от структуры базы данных
	return nil
}

// SetBotInstance устанавливает экземпляр бота
func (b *BuildingMonitor) SetBotInstance(botInstance interface{}) {
	b.botInstance = botInstance
}

// IsTrackingActive проверяет, активен ли трекинг для пользователя
func (b *BuildingMonitor) IsTrackingActive(telegramID int64, playerTag string) (bool, error) {
	trackers, err := b.dbService.GetUserBuildingTrackers(telegramID)
	if err != nil {
		return false, err
	}

	for _, tracker := range trackers {
		if tracker.PlayerTag == playerTag && tracker.IsActive {
			return true, nil
		}
	}

	return false, nil
}
